pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorService.java

```diff
@@ -59,30 +62,77 @@ public class SelectionOperatorService {
   private final QueryContext _queryContext;
   private final DataSchema _dataSchema;
   private final int[] _columnIndices;
-  private final int _offset;
+  private int _offset;
+  private final int _limit;
   private final int _numRowsToKeep;
-  private final PriorityQueue<Object[]> _rows;
+  // NOTE: this pq is only useful for server version < 1.3.0
+  private final PriorityQueue<Object[]> _heapSortRows;
+  private final PriorityQueue<MergeItem> _mergeSortRows;
+  private final Comparator<Object[]> _comparator;
+  private final Comparator<MergeItem> _mergeComparator;
+
+  /** Util class used for n-way merge */
+  private static class MergeItem {
+    protected final Object[] _row;
+    protected final int _dataTableId;
+
+    public MergeItem(Object[] row, int dataTableId) {
```

Comment:
For private inner helper class
```suggestion
   final Object[] _row;
   final int _dataTableId;

   MergeItem(Object[] row, int dataTableId) {
```